sentry-android-core/src/main/java/io/sentry/android/core/AndroidTraceTransactionListener.java

```diff
@@ -0,0 +1,150 @@
+package io.sentry.android.core;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+import android.os.Build;
+import android.os.Debug;
+import io.sentry.HubAdapter;
+import io.sentry.IHub;
+import io.sentry.ITransaction;
+import io.sentry.ITransactionListener;
+import io.sentry.SentryEnvelope;
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.protocol.SentryId;
+import io.sentry.util.FileUtils;
+import io.sentry.util.Objects;
+import io.sentry.util.SentryExecutors;
+import java.io.File;
+import java.io.IOException;
+import java.util.UUID;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class AndroidTraceTransactionListener implements ITransactionListener {
+
+  /**
+   * This appears to correspond to the buffer size of the data part of the file, excluding the key
+   * part. Once the buffer is full, new records are ignored, but the resulting trace file will be
+   * valid.
+   *
+   * <p>30 second traces can require a buffer of a few MB. 8MB is the default buffer size for
+   * [Debug.startMethodTracingSampling], but 3 should be enough for most cases. We can adjust this
+   * in the future if we notice that traces are being truncated in some applications.
+   */
+  private static final int BUFFER_SIZE_BYTES = 3_000_000;
+
+  private final IHub hub;
+
+  private @Nullable File traceFile = null;
+  private @Nullable File traceFilesDir = null;
+  private boolean startedMethodTracing = false;
+  private @Nullable ITransaction activeTransaction = null;
+
+  private @NotNull final SentryOptions options;
+
+  public AndroidTraceTransactionListener() {
+    this(HubAdapter.getInstance());
+  }
+
+  public AndroidTraceTransactionListener(IHub hub) {
```

Comment:
please always use `final ` `@NotNull` or `@Nullable` for ctor params, method params etc otherwise static code analysis get unhappy